### PR TITLE
Fixes polynomial solve with GoldenRatio and TribonacciConstant

### LINKED ISSUE #11553: Polynomial solve with GoldenRatio causes Traceback
Found during issue 11538 investigation. The GoldenRatio::NumberSymbol is the only NumberSymbol currently treated as an algebraic expression. However when solve is called on the GoldenRatio a Traceback occurs where GoldenRatio is not being treated algebraically. 

```
Python 2.7.12 (v2.7.12:d33e0cf91556, Jun 27 2016, 15:19:22) [MSC v.1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import sympy
>>> x,y=sympy.symbols('x y')
>>> e1=x+y+1
>>> e2=x+sympy.GoldenRatio
>>> sympy.solve([e1,e2],x,y)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\solvers\solvers.py", line 1053, in solve
    solution = _solve_system(f, symbols, **flags)
  File "sympy\solvers\solvers.py", line 1660, in _solve_system
    poly = g.as_poly(*symbols, extension=True)
  File "sympy\core\basic.py", line 699, in as_poly
    poly = Poly(self, *gens, **args)
  File "sympy\polys\polytools.py", line 89, in __new__
    return cls._from_expr(rep, opt)
  File "sympy\polys\polytools.py", line 200, in _from_expr
    return cls._from_dict(rep, opt)
  File "sympy\polys\polytools.py", line 144, in _from_dict
    domain, rep = construct_domain(rep, opt=opt)
  File "sympy\polys\constructor.py", line 238, in construct_domain
    result = _construct_simple(coeffs, opt)
  File "sympy\polys\constructor.py", line 45, in _construct_simple
    domain, result = _construct_algebraic(coeffs, opt)
  File "sympy\polys\constructor.py", line 93, in _construct_algebraic
    g, span, H = primitive_element(exts, ex=True, polys=True)
  File "sympy\polys\numberfields.py", line 857, in primitive_element
    f = minpoly(ext, y)
  File "sympy\polys\numberfields.py", line 657, in minimal_polynomial
    result = _minpoly_compose(ex, x, dom)
  File "sympy\polys\numberfields.py", line 578, in _minpoly_compose
    raise NotAlgebraic("%s doesn't seem to be an algebraic element" % ex)
sympy.polys.polyerrors.NotAlgebraic: GoldenRatio doesn't seem to be an algebraic element
```

### PR DESCRIPTION
**Pull request by Smit-create**
Fixes polynomial solve with GoldenRatio and TribonacciConstant
#### References to other Issues or PRs

Fixes #11553

#### Brief description of what is fixed or changed
Before fixing:
```
>>> from sympy import *
>>> x,y = symbols('x y')
>>> eq1 = x + y + 1
>>> eq2 = x + GoldenRatio
>>> solve([eq1, eq2], x, y)
Traceback (most recent call last):
   ...
sympy.polys.polyerrors.NotAlgebraic: GoldenRatio doesn't seem to be an algebraic element
```

After fixing:
```
>>> from sympy import *
>>> x,y = symbols('x y')
>>> eq1 = x + y + 1
>>> eq2 = x + GoldenRatio
>>> solve([eq1, eq2], x, y)
{x: -GoldenRatio, y: -1 + GoldenRatio}
```

#### Other comments
GoldenRatio is the number symbol treated as an algebraic expression as `(x**2 - x - 1)
`
TribonacciConstant is the number symbol treated as an algebraic expression as `(x**3 - x**2 - x - 1)
`
#### Release Notes

* polys
  * Added check for expressions containing GoldenRatio and TribonacciConstant

### CODE REVIEW

File: sympy/polys/numberfields.py [hunk 1]
```diff
@@ -517,6 +517,8 @@ def _minpoly_compose(ex, x, dom):
     if ex is I:
         _, factors = factor_list(x**2 + 1, x, domain=dom)
         return x**2 + 1 if len(factors) == 1 else x - I
+    if ex is GoldenRatio:
+        return x**2 - x - 1
```
Comment (asmeurer) [inline sympy/polys/numberfields.py]: I think we need the same factorization check as above with `I` to see if phi is already included in the domain.

Comment (jksuom) [inline sympy/polys/numberfields.py]: Maybe this should be handled in the same way as `I`: It `GoldenRatio` is an element of `dom`, then `x**2 - x - 1` splits into two linear factors over `dom` and the minimal polynomial is one them.

File: sympy/polys/numberfields.py [hunk 2]
```diff
@@ -518,7 +518,8 @@ def _minpoly_compose(ex, x, dom):
         _, factors = factor_list(x**2 + 1, x, domain=dom)
         return x**2 + 1 if len(factors) == 1 else x - I
     if ex is GoldenRatio:
-        return x**2 - x - 1
+        _, factors = factor_list(x**2 - x - 1, x, domain=dom)
+        return x**2 - x - 1 if len(factors) == 1 else x - GoldenRatio
```
Comment (jksuom) [inline sympy/polys/numberfields.py]: The result should probably not contain `GoldenRatio` (danger of circularity). Instead, one should choose one of the factors using `_choose_factor` with argument `(1 + sqrt(5))/2`.

**Comment by sylee957** [comment]
It also happens with `TribonacciConstant`. Maybe you should also fix this in similar way.
But I would have some questions like, why sympy puts such trivial algebraic constants as singleton classes.

**Comment by jksuom** [comment]
> why sympy puts such trivial algebraic constants as singleton classes.

I have also wondered. It should be investigated if they could be algebraic numbers.

File: sympy/polys/numberfields.py [hunk 3]
```diff
@@ -517,9 +517,22 @@ def _minpoly_compose(ex, x, dom):
...
+            return x**2 - x - 1
+        else:
+            return _choose_factor(factors, x, (1 + sqrt(5))/2, dom=dom)
+
+    if ex is TribonacciConstant:
+        _, factors = factor_list(x**3 - x**2 - x - 1, x, domain=dom)
+        if len(factors) == 1:
+            return x**3 - x**2 - x - 1
+        else:
+            fac = 1/3 + 4/(9*(sqrt(33)/9 + 19/27)**(1/3)) + (sqrt(33)/9 + 19/27)**(1/3)
```
Comment (jksuom) [inline sympy/polys/numberfields.py]: `cbrt()` could be used here.

**Comment by jksuom** [comment]
Can you add tests for `minimal_polynomial(e, x)` where `e` is `GoldenRatio` or `TribonacciConstant` or and expression containing them. Also for `minimal_polynomial(e, x, domain=K)` where `K = QQ.algebraic_field(sqrt(5))` and `e` has `GoldenRatio`.

**Comment by Smit-create** [comment]
Yes, would add them.

**Comment by jksuom** [comment]
Thanks, this is ready to go.

**Comment by Smit-create** [comment]
Thanks @jksuom

**Pull request merged**